File: Sources/windows/yage/capabilities/bounce_on_lateral_collisions.py
import math
from typing import List
from yage.models.capability import Capability
from yage.models.collisions import Collision, CollisionSide
from yage.models.entity_state import EntityState
from yage.utils.geometry import Vector


class BounceOnLateralCollisions(Capability):
    def do_update(self, collisions: List[Collision], time: float):
        if self.subject.is_ephemeral or self.subject.state != EntityState.MOVE:
            return
        angle = self.bouncing_angle(self.subject.direction.radians, collisions)
        if angle is None:
            return
        self.subject.direction = Vector.from_radians(angle)

    def bouncing_angle(self, current_angle: float, collisions: List[Collision]) -> float:
        target_side = self.target_side()
        valid_collisions = [c for c in collisions if c.other.is_static]
        if not target_side:
            return None
        if not _contains_overlap_on_side(valid_collisions, target_side):
            return None
        if _contains_overlap_on_side(valid_collisions, target_side.opposite):
            return None
        return math.pi - current_angle

    def target_side(self) -> CollisionSide:
        direction = self.subject.direction.dx
        if direction < -0.0001:
            return CollisionSide.LEFT
        if direction > 0.0001:
            return CollisionSide.RIGHT
        return None


def _contains_overlap_on_side(collisions, target_side) -> bool:
    valid_collisions = filter(
        lambda c: not c.is_ephemeral and c.is_overlapping, collisions)
    return any(target_side in c.sides() for c in valid_collisions)

File: Sources/macOS/Sources/Yage/Capabilities/BounceOnLateralCollisions.swift
import SwiftUI

public class BounceOnLateralCollisions: Capability {
    public var customCollisionsFilter: ((Collision) -> Bool)?
    
    override public func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        guard let subject, !subject.isEphemeral, subject.state == .move else { return }
        guard let angle = bouncingAngle(from: subject.direction.radians, with: collisions) else { return }
        subject.direction = CGVector(radians: angle)
    }

    func bouncingAngle(from currentAngle: CGFloat, with collisions: Collisions) -> CGFloat? {
        guard let targetSide = targetSide() else { return nil }
        let validCollisions = collisions.filter { isValidCollision($0) }
        guard validCollisions.contains(overlapOnSide: targetSide) else { return nil }
        guard !validCollisions.contains(overlapOnSide: targetSide.opposite) else { return nil }
        return CGFloat.pi - currentAngle
    }
    
    private func isValidCollision(_ collision: Collision) -> Bool {
        if let filter = customCollisionsFilter {
            return filter(collision)
        } else {
            return collision.other?.isStatic == true
        }
    }

    private func targetSide() -> Collision.Side? {
        guard let direction = subject?.direction.dx else { return nil }
        if direction < -0.0001 { return .left }
        if direction > 0.0001 { return .right }
        return nil
    }
}

private extension Collisions {
    func contains(overlapOnSide targetSide: Collision.Side) -> Bool {
        contains {
            guard !$0.isEphemeral else { return false }
            guard $0.isOverlapping else { return false }
            guard $0.sides().contains(targetSide) else { return false }
            return true
        }
    }
}

private extension Collision.Side {
    var opposite: Collision.Side {
        switch self {
        case .top: return .bottom
        case .bottom: return .top
        case .left: return .right
        case .right: return .left
        }
    }
}

public extension Entity {
    func setBounceOnLateralCollisions(enabled: Bool) {
        capability(for: BounceOnLateralCollisions.self)?.isEnabled = enabled
    }
}
